### PR TITLE
fix(metrics): stop using user-agent string in flow id check

### PR DESCRIPTION
**Pull request by philbooth**
We're tweaking the flow id validation a little bit, so this PR introduces a dual check. First it tries to check the new way then, if that fails, we try again the old way. In the next train, we can remove the old way entirely and update the tests to match.

Only way to test this really is to make sure flow events are emitted when it is run against the content server in `master` and the content server in `pb/flow-sig-ua`, which I'll be opening a PR for presently.

@mozilla/fxa-devs r?

### CODE REVIEW

File: lib/metrics/context.js [hunk 1]
```diff
@@ -197,9 +197,14 @@ module.exports = function (log, config) {
     // share state between content-server and auth-server.
     const flowSignature = metadata.flowId.substr(FLOW_ID_LENGTH / 2, FLOW_ID_LENGTH)
     const flowSignatureBytes = Buffer.from(flowSignature, 'hex')
-    const expectedSignatureBytes = calculateFlowSignatureBytes(this, metadata)
+    let expectedSignatureBytes = calculateFlowSignatureBytes(metadata)
     if (! bufferEqualConstantTime(flowSignatureBytes, expectedSignatureBytes)) {
-      return logInvalidContext(this, 'invalid signature')
+      // HACK: We're transitioning between flow id recipes so, just for one train,
+      //       fall back to trying the old way if the preceding check failed.
```
Comment (rfk) [inline lib/metrics/context.js]: 👍 nice

**Comment by philbooth** [comment]
> ...testing this with auth `master` and content `master` creating an account, I was expecting no flow events to be emitted when I verified on another browser. However, the `account.verified` events got emitted.

@vbudhram that is the expected behaviour for `account.verified` and `account.confirmed` events because `/recovery_email/verify_code` doesn't take a `metricsContext` payload, so it doesn't call `request.validateMetricsContext()`. Instead it reads the data from memcached and there's no need to validate the flow id at that point.

So for those flows, it's only content server events that are affected by the user-agent thing. For instance, in `master` you should only see a `flow.signup.sms.view` event emitted from the content server after verifying an account. Whereas in `pb/flow-sig-ua` you should see that preceded by `flow.signup-verified.view` and a bunch of performance events:

```
{"event":"flow.signup-verified.view","flow_id":"af7f62ceff442665a14258719a9e8e153c7ca7e8b8f3139a784cf6257bfd30a3","flow_time":298291,"hostname":"Phils-MBP","locale":"en","op":"flowEvent","pid":78214,"time":"2018-04-12T08:04:52.895Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:61.0) Gecko/20100101 Firefox/61.0","v":1,"context":"fx_sync","entrypoint":"menupanel","service":"sync"}{"event":"flow.performance.other","flow_id":"af7f62ceff442665a14258719a9e8e153c7ca7e8b8f3139a784cf6257bfd30a3","flow_time":298294,"hostname":"Phils-MBP","locale":"en","op":"flowEvent","pid":78214,"time":"2018-04-12T08:04:52.898Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:61.0) Gecko/20100101 Firefox/61.0","v":1,"context":"fx_sync","entrypoint":"menupanel","service":"sync"}
{"event":"flow.performance.other.network","flow_id":"af7f62ceff442665a14258719a9e8e153c7ca7e8b8f3139a784cf6257bfd30a3","flow_time":2,"hostname":"Phils-MBP","locale":"en","op":"flowEvent","pid":78214,"time":"2018-04-12T07:59:54.606Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:61.0) Gecko/20100101 Firefox/61.0","v":1,"context":"fx_sync","entrypoint":"menupanel","service":"sync"}
{"event":"flow.performance.other.server","flow_id":"af7f62ceff442665a14258719a9e8e153c7ca7e8b8f3139a784cf6257bfd30a3","flow_time":113,"hostname":"Phils-MBP","locale":"en","op":"flowEvent","pid":78214,"time":"2018-04-12T07:59:54.717Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:61.0) Gecko/20100101 Firefox/61.0","v":1,"context":"fx_sync","entrypoint":"menupanel","service":"sync"}
{"event":"flow.performance.other.client","flow_id":"af7f62ceff442665a14258719a9e8e153c7ca7e8b8f3139a784cf6257bfd30a3","flow_time":374,"hostname":"Phils-MBP","locale":"en","op":"flowEvent","pid":78214,"time":"2018-04-12T07:59:54.978Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:61.0) Gecko/20100101 Firefox/61.0","v":1,"context":"fx_sync","entrypoint":"menupanel","service":"sync"}
{"event":"flow.performance.other.connectStart","flow_id":"af7f62ceff442665a14258719a9e8e153c7ca7e8b8f3139a784cf6257bfd30a3","flow_time":58,"hostname":"Phils-MBP","locale":"en","op":"flowEvent","pid":78214,"time":"2018-04-12T07:59:54.662Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:61.0) Gecko/20100101 Firefox/61.0","v":1,"context":"fx_sync","entrypoint":"menupanel","service":"sync"}
{"event":"flow.performance.other.domainLookupEnd","flow_id":"af7f62ceff442665a14258719a9e8e153c7ca7e8b8f3139a784cf6257bfd30a3","flow_time":58,"hostname":"Phils-MBP","locale":"en","op":"flowEvent","pid":78214,"time":"2018-04-12T07:59:54.662Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:61.0) Gecko/20100101 Firefox/61.0","v":1,"context":"fx_sync","entrypoint":"menupanel","service":"sync"}
{"event":"flow.performance.other.requestStart","flow_id":"af7f62ceff442665a14258719a9e8e153c7ca7e8b8f3139a784cf6257bfd30a3","flow_time":61,"hostname":"Phils-MBP","locale":"en","op":"flowEvent","pid":78214,"time":"2018-04-12T07:59:54.665Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:61.0) Gecko/20100101 Firefox/61.0","v":1,"context":"fx_sync","entrypoint":"menupanel","service":"sync"}
{"event":"flow.signup.sms.view","flow_id":"af7f62ceff442665a14258719a9e8e153c7ca7e8b8f3139a784cf6257bfd30a3","flow_time":300589,"hostname":"Phils-MBP","locale":"en","op":"flowEvent","pid":78214,"time":"2018-04-12T08:04:55.193Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
```

However, if you complete the `account.reset` flow in a different browser, you should see the problem manifested in the auth server too, because those endpoints take a `metricsContext` payload and call `request.validateMetricsContext()`.

Sorry for not making all this clear in the other issue, I realise a blanket "no flow events are emitted" statement sounded melodramatic in hindsight. Still worth fixing though I reckon!

**Comment by philbooth** [comment]
@shane-tomlinson huh, that's weird. The two browser instances have different user-agent strings, yeah? And you have the content server `pb/flow-sig-ua` branch checked out?

Anyway, these are the exact steps I'm doing:

1. Filter the content server logs with `./pm2 logs 3 | grep flowEvent`, for easier debugging.
2. Sign up for a Sync account in browser "A" (happens to be Fx 58 in my case, but whatevs).
3. Click "Save settings" on CWTS.
4. Hit enter a bunch of times in the content server log window, so we know where we are in the event flow.
5. Verify the account in browser "B" (happens to be Fx 61 in my case, but whatevs).
6. Blur both browser windows so that metrics flush.

At this point, if I have the two `master` branches checked out, I only see one flow event in the content server:

```
3|content- | {"event":"flow.signin.view","flow_id":"e2f5005eb3035441b0a2391e8ffb7dd9a502121a846bb42fe4cef13819e283fb","flow_time":63100,"hostname":"Phils-MBP","locale":"en","op":"flowEvent","pid":89492,"time":"2018-04-12T08:46:50.680Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"preferences","service":"sync"}
```

But if I have `pb/flow-sig-ua` I see these:

```
3|content- | {"event":"flow.signup-verified.view","flow_id":"c4baf7f00d82fbe844fef54484b41fca5b2512a7c5ed5b03bb1cd62c82616ae1","flow_time":27772,"hostname":"Phils-MBP","locale":"en","op":"flowEvent","pid":16186,"time":"2018-04-12T11:39:28.531Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:61.0) Gecko/20100101 Firefox/61.0","v":1,"context":"fx_sync","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.performance.other","flow_id":"c4baf7f00d82fbe844fef54484b41fca5b2512a7c5ed5b03bb1cd62c82616ae1","flow_time":27774,"hostname":"Phils-MBP","locale":"en","op":"flowEvent","pid":16186,"time":"2018-04-12T11:39:28.533Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:61.0) Gecko/20100101 Firefox/61.0","v":1,"context":"fx_sync","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.performance.other.network","flow_id":"c4baf7f00d82fbe844fef54484b41fca5b2512a7c5ed5b03bb1cd62c82616ae1","flow_time":1,"hostname":"Phils-MBP","locale":"en","op":"flowEvent","pid":16186,"time":"2018-04-12T11:39:00.760Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:61.0) Gecko/20100101 Firefox/61.0","v":1,"context":"fx_sync","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.performance.other.server","flow_id":"c4baf7f00d82fbe844fef54484b41fca5b2512a7c5ed5b03bb1cd62c82616ae1","flow_time":63,"hostname":"Phils-MBP","locale":"en","op":"flowEvent","pid":16186,"time":"2018-04-12T11:39:00.822Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:61.0) Gecko/20100101 Firefox/61.0","v":1,"context":"fx_sync","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.performance.other.client","flow_id":"c4baf7f00d82fbe844fef54484b41fca5b2512a7c5ed5b03bb1cd62c82616ae1","flow_time":321,"hostname":"Phils-MBP","locale":"en","op":"flowEvent","pid":16186,"time":"2018-04-12T11:39:01.080Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:61.0) Gecko/20100101 Firefox/61.0","v":1,"context":"fx_sync","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.performance.other.connectStart","flow_id":"c4baf7f00d82fbe844fef54484b41fca5b2512a7c5ed5b03bb1cd62c82616ae1","flow_time":61,"hostname":"Phils-MBP","locale":"en","op":"flowEvent","pid":16186,"time":"2018-04-12T11:39:00.820Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:61.0) Gecko/20100101 Firefox/61.0","v":1,"context":"fx_sync","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.performance.other.domainLookupEnd","flow_id":"c4baf7f00d82fbe844fef54484b41fca5b2512a7c5ed5b03bb1cd62c82616ae1","flow_time":61,"hostname":"Phils-MBP","locale":"en","op":"flowEvent","pid":16186,"time":"2018-04-12T11:39:00.820Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:61.0) Gecko/20100101 Firefox/61.0","v":1,"context":"fx_sync","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.performance.other.requestStart","flow_id":"c4baf7f00d82fbe844fef54484b41fca5b2512a7c5ed5b03bb1cd62c82616ae1","flow_time":64,"hostname":"Phils-MBP","locale":"en","op":"flowEvent","pid":16186,"time":"2018-04-12T11:39:00.823Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:61.0) Gecko/20100101 Firefox/61.0","v":1,"context":"fx_sync","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.signup.sms.view","flow_id":"c4baf7f00d82fbe844fef54484b41fca5b2512a7c5ed5b03bb1cd62c82616ae1","flow_time":27890,"hostname":"Phils-MBP","locale":"en","op":"flowEvent","pid":16186,"time":"2018-04-12T11:39:28.649Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
```

**Comment by philbooth** [comment]
(or you can also test with auth server `account.reset` event, if you want to try that flow)

**Comment by vbudhram** [comment]
@philbooth Thanks for the STR! I was looking at auth-server flow events and not content-server >.<

After doing this, I was able to see the changes you described. With `pb/flow-sig-ua` I get these on chrome when verifying a new account
```
3|content- | {"event":"flow.sms.ineligible.no_session","flow_id":"4d79d135a51fba728e5385a6dc8331f875216cc6096acba3e3343c60d8a23f45","flow_time":44779,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":38758,"time":"2018-04-12T13:50:48.499Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.181 Safari/537.36","v":1,"context":"fx_sync","entrypoint":"menupanel","service":"sync","utm_campaign":"fx-welcome","utm_content":"fx-activate","utm_medium":"email","utm_source":"email"}
3|content- | {"event":"flow.connect-another-device.view","flow_id":"4d79d135a51fba728e5385a6dc8331f875216cc6096acba3e3343c60d8a23f45","flow_time":44800,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":38758,"time":"2018-04-12T13:50:48.520Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.181 Safari/537.36","v":1,"context":"fx_sync","entrypoint":"menupanel","service":"sync","utm_campaign":"fx-welcome","utm_content":"fx-activate","utm_medium":"email","utm_source":"email"}
3|content- | {"event":"flow.signedin.false","flow_id":"4d79d135a51fba728e5385a6dc8331f875216cc6096acba3e3343c60d8a23f45","flow_time":44800,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":38758,"time":"2018-04-12T13:50:48.520Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.181 Safari/537.36","v":1,"context":"fx_sync","entrypoint":"menupanel","service":"sync","utm_campaign":"fx-welcome","utm_content":"fx-activate","utm_medium":"email","utm_source":"email"}
```

These flowEvents make sense to me. I also noticed that the utm params are being emitted, maybe a partial fix for https://github.com/mozilla/fxa-amplitude-send/issues/44?

**Pull request merged**